DDP-5658 support revisioned templates

yufengwng opened the pull request:
## Context

See DDP-5658. Database supports storing revisioned templates, but our queries are not accounting for revisioned "template variable substitutions". This PR refactors the code to support such revisioned templates. This will make it easier for studies to create new revisions of activities that requires only simple word change to templates, as in DDP-5583.

Key concepts:
* Templates contain a list of variables, and each variable have a list of translation substitution texts.
* These variable substitutions can be revisioned using time-based approach (e.g. start date, end date).
* So when querying for revisioned templates, we'll need a timestamp to pinpoint the revision.
* When querying for activity definitions, we use the activity's version start time as timestamp.
* When querying for activity instances, we use the instance's created_at time as timestamp.

I have ran several studies locally, and tested fetching of activity data for various revisions. It looks fine for me locally as far as I can tell. When this is merged, I'll ask for Kiara's help to thoroughly regression test this.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [x] :sweat_smile: There might be some issues here

## How do we demo these changes?

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!

yufengwng committed:
DDP-5658 support revisioned templates

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivity.java:
```diff
@@ -180,10 +180,6 @@ int updateMaxInstancesPerUserById(
             @Bind("maxInstances") Integer maxInstances
     );
 
-    default List<ActivityValidationDto> findValidationsById(long activityId, long languageCodeId) {
```
This method is not used, so I removed it.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivityValidation.java:
```diff
@@ -58,17 +56,6 @@ int _insertActivityValidation(
             @Bind("errorMessageTemplateId") long errorMessageTemplateId
     );
 
-    default List<ActivityValidationDto> _findByActivityIdTranslated(long activityId, long isoLanguageCode) {
```
This method is not used.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java:
```diff
@@ -370,7 +371,10 @@ public MailAddress verifyAddress(MailAddress address) throws AddressVerification
+                        // Query using the current time to get the latest template.
+                        msg = handle.attach(TemplateDao.class)
+                                .loadTemplateByIdAndTimestamp(warningTmplId, Instant.now().toEpochMilli())
+                                .render(langCode);
```
NPE probability: loadTemplateByIdAndTimestamp can return null

yufengwng replied:
We checked that `warningTmplId != null` so we're expecting to get back an object here.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/util/StatisticsUtil.java:
```diff
@@ -84,7 +85,10 @@ private static StatisticsFigure generateStatisticsItem(Handle handle,
+
+            // As we see above, we're querying the latest question and picklist options,
+            // so we use the current time here to render the latest templates.
+            renderer.bulkRenderAndApply(handle, options, style, langId, Instant.now().toEpochMilli());
```
Does it make sense to create util method: `long epochMilli() {return Instant.now().toEpochMilli();}` ?
Then code could be a little bit compact.

yufengwng replied:
I think the way it is now isn't too bad.

evgeniipr commented on pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:
```diff
@@ -33,8 +41,12 @@ select t.template_id,
-<select_all_templates()> where t.template_id in (<templateIds>)
+queryTemplatesByIdsAndTimestamp(templateIds) ::= <<
+<select_all_templates()>
+ where t.template_id in (<templateIds>)
```
Is it needed slashes before brackets: `<templateIds>` ?

yufengwng replied:
For this particular query, I found that we don't need slashes for escaping here.

evgeniipr commented on pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:
```diff
@@ -8,11 +8,19 @@ from
+  left join revision as i18n_ts_def_rev on
+    i18n_ts_def_rev.revision_id = i18n_ts_def.revision_id
 where
   tv.template_id in (\<templateIds\>) and
```
I found out weird thing: slash before closing bracket seems not necessary: `\<templateIds>`.
I'm not 100% sure but when I tried to enable SQL logging for JDBI I encountered failures (fatal errors) with logging SQLs with params like this: in (\<templateIds\>). I experimented different ways and eventually made it working by removing the slash before closing bracket.
Btw, the logging can be activated in JDBI version 3.18.0 (in pepper now used 3.1.1) and it needs to add the following block of code in TransactionWrapper.java in constructor:
```
 SqlLogger sqlLogger = new SqlLogger() {
            @Override
            public void logAfterExecution(StatementContext context) {
                LOG.info("sql {}, parameters {}, timeTaken {} ms", context.getRenderedSql(),
                        context.getBinding().toString(), context.getElapsedTime(ChronoUnit.MILLIS));
            }
        };
        jdbi.setSqlLogger(sqlLogger);
```

yufengwng replied:
Yeah, whether we need slashes or not depends on a few things like, what the query is, if we're binding a list or not, and how the query is used in the code.

yufengwng replied:
Since this existing query has slashes, I'm keeping it as-is.

yufengwng commented:
I found an issue with revisioned templates and activity-level validations while doing some local testing. I'm working on a fix.

yufengwng committed:
DDP-5658 fix revisioning issue with Brain study activity validations

yufengwng committed:
Merge branch 'develop' into ddp-5658-revisioned-templates

yufengwng commented:
@evgeniipr @MocanaAtBroad The issue is now fixed. See `BrainFixPrequalValidations` for more details. Basically, the Brain study configuration had some malformed revisions for the Prequal's activity validations. I added code to fix up those revision data.
